### PR TITLE
{Network} `az network nsg`: Refactor network security group

### PR DESCRIPTION
**Pull request by necusjz**
**Related command**

az network nsg
az network nsg rule

**Description**

Fixes: https://github.com/Azure/azure-cli/issues/24910
Fixes: https://github.com/Azure/azure-cli/issues/24939

**Testing Guide**

**History Notes**

[Component Name 1] BREAKING CHANGE: `az command a`: Make some customer-facing breaking change
[Component Name 2] `az command b`: Add some customer-facing feature

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Network

File: src/azure-cli/azure/cli/command_modules/network/aaz/latest/network/nsg/rule/_list.py [hunk 1]
```diff
@@ -43,9 +43,9 @@ def _build_arguments_schema(cls, *args, **kwargs):
         # define Arg Group ""
 
         _args_schema = cls._args_schema
-        _args_schema.name = AAZStrArg(
-            options=["-n", "--name"],
-            help="The name of the network security group.",
+        _args_schema.nsg_name = AAZStrArg(
+            options=["--nsg-name"],
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/aaz/latest/network/nsg/rule/_list.py:47]: Is it a breaking change?

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 2]
```diff
@@ -5186,169 +5180,157 @@ def remove_nic_ip_config_inbound_nat_rule(
...
-        "location": location,
-        "tags": tags
-    })
+def _handle_plural_or_singular(args, plural_name, singular_name):
+    values = getattr(args, plural_name)
+    if not has_value(values):
+        return
 
+    setattr(args, plural_name, values if len(values) > 1 else None)
+    setattr(args, singular_name, values[0] if len(values) == 1 else None)
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/custom.py:5189]: Can values == None?

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 3]
```diff
@@ -5186,149 +5180,150 @@ def remove_nic_ip_config_inbound_nat_rule(
...
-            help="Space-separated list of application security group names or supports one application security group name or ID. Use null to detach it.",
-            fmt=EmptyListArgFormat()
+            fmt=EmptyListArgFormat(),
         )
         args_schema.destination_asgs.Element = AAZResourceIdArg(
             nullable=True,
-            fmt=EmptyResourceIdArgFormat(
-                template="/subscriptions/{subscription}/resourceGroups/{resource_group}/providers/Microsoft.Network/applicationSecurityGroups/{}"
-            )
+            fmt=AAZResourceIdArgFormat(
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/custom.py:5277]: Why it's changed from `EmptyResourceIdArgFormat` to `AAZResourceIdArgFormat`. It's a breaking change.

**Review by kairu-ms — APPROVED** [review]

**Pull request merged**